Sources/SourceKitLSP/TestDiscovery.swift

```diff
@@ -462,24 +485,131 @@ extension TestItem {
+  /// If we think of the enclosing class or struct as the test suite, then extensions on that class or struct should be
+  /// additions to that suite, just like extensions on types are, from the user's perspective, transparently added to
+  /// their type.
+  ///
+  /// This method walks the `AnnotatedTestItem` tree produced by the test scanners and merges in the tests defined in
+  /// extensions into the final `TestItem`s that represent the type definition.
+  ///
+  /// This causes extensions to be merged into their type's definition if the type's definition exists in the list of
+  /// test items. If the type's definition is not a test item in this collection, the first extension of that type will
+  /// be used as the primary test location.
+  ///
+  /// For example if there are two files
+  ///
+  /// FileA.swift
+  /// ```swift
+  /// @Suite struct MyTests {
+  ///   @Test func oneIsTwo {}
+  /// }
+  /// ```
+  ///
+  /// FileB.swift
+  /// ```swift
+  /// extension MyTests {
+  ///   @Test func twoIsThree() {}
+  /// }
+  /// ```
+  ///
+  /// Then `workspace/tests` will return
+  /// - `MyTests` (FileA.swift:1)
+  ///   - `oneIsTwo`
+  ///   - `twoIsThree`
+  ///
+  /// And `textDocument/tests` for FileB.swift will return
+  /// - `MyTests` (FileB.swift:1)
+  ///   - `twoIsThree`
+  ///
+  /// A node's parent is identified by the node's ID with the last component dropped.
+  func mergingTestsInExtensions() -> [TestItem] {
+    var itemDict: [String: AnnotatedTestItem] = [:]
+    for item in self {
+      let id = item.testItem.id
+      if var rootItem = itemDict[id] {
+        // If we've encountered an extension first, and this is the
+        // type declaration, then use the type declaration TestItem
+        // as the root item.
+        if rootItem.isExtension && !item.isExtension {
+          var newItem = item
+          newItem.testItem.children += rootItem.testItem.children
+          rootItem = newItem
+        } else {
+          rootItem.testItem.children += item.testItem.children
+        }
+
+        itemDict[id] = rootItem
+      } else {
+        itemDict[id] = item
+      }
+    }
+
+    if itemDict.isEmpty {
+      return []
+    }
+
+    var mergedIds = Set<String>()
+    for item in self {
+      let id = item.testItem.id
+      let parentID = id.components(separatedBy: "/").dropLast().joined(separator: "/")
+      // If the parent exists, add the current item to its children and remove it from the root
+      if var parent = itemDict[parentID] {
+        parent.testItem.children.append(item.testItem)
+        mergedIds.insert(parent.testItem.id)
+        itemDict[parent.testItem.id] = parent
+        itemDict[id] = nil
+      }
+    }
+
+    // Filter out the items that have been merged into their parents, sorting the tests by location.
+    // TestItems not in extensions should be priotitized first.
+    let sortedItems = itemDict.values
+      .compactMap { $0 }
```

Comment:
I don’t think the `compactMap` is doing anything. `itemDict.values` is has element `AnnotatedTestItem`, which is never `nil`.

Should also update the comment above accordingly when `compactMap` is removed.

Maybe you got confused because you did `itemDict[id] = nil`, which removes the key from the dictionary in Swift instead of setting the value to `nil`.